pkg/actuators/awsplacementgroup/controller.go

```diff
@@ -3,91 +3,622 @@ package awsplacementgroup
 		For(&machinev1.AWSPlacementGroup{}).
+		Watches(&source.Kind{Type: &machinev1beta1.Machine{}}, handler.EnqueueRequestsFromMapFunc(machineToAWSPlacementGroup(r))).
 		WithOptions(options).
-		Build(r)
-
-	if err != nil {
+		Complete(r); err != nil {
 		return fmt.Errorf("failed setting up with a controller manager: %w", err)
 	}
 
 	r.recorder = mgr.GetEventRecorderFor("awsplacementgroup-controller")
 	r.scheme = mgr.GetScheme()
+
 	return nil
 }
 
-// Reconcile implements controller runtime Reconciler interface.
-func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	logger := r.Log.WithValues("name", req.Name, "namespace", req.Namespace)
-	logger.V(3).Info("Reconciling")
+// Reconcile implements controller runtime AWSPlacementGroupReconciler interface.
+func (r *AWSPlacementGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Log.WithValues("namespace", req.Namespace, "name", req.Name)
+	logger.V(3).Info("Reconciling AWSPlacementGroup")
 
 	awsPlacementGroup := &machinev1.AWSPlacementGroup{}
 	if err := r.Client.Get(ctx, req.NamespacedName, awsPlacementGroup); err != nil {
 		if apierrors.IsNotFound(err) {
-			// Object not found, return. Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
+			// Object not found.
+			// Return without requeue as created objects are automatically garbage collected.
 			return ctrl.Result{}, nil
 		}
-		// Error reading the object - requeue the request.
+		// For any other type of error, requeue immediately.
 		return ctrl.Result{}, err
 	}
 
-	// Ignore deleted AWSPlacementGroups, this can happen when foregroundDeletion
-	// is enabled
-	if !awsPlacementGroup.DeletionTimestamp.IsZero() {
+	if err := validateAWSPlacementGroup(awsPlacementGroup); err != nil {
+		logger.Error(err, fmt.Sprintf("AWSPlacementGroup %s is invalid", awsPlacementGroup.Name))
+		// Return without erroring to avoid requeue.
+		// The object shouldn't be requeued until it has been modified and is ready to be validated again,
+		// as it is invalid.
 		return ctrl.Result{}, nil
 	}
 
+	// Get the Infrastructure object.
+	infra := &configv1.Infrastructure{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: awsclient.GlobalInfrastuctureName}, infra); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if r.Infrastructure == nil {
+		r.Infrastructure = infra
+	}
+
+	// Check if the CredentialsSecret is defined,
+	// then obtain its name for later use.
+	credentialsSecretName := ""
+	if awsPlacementGroup.Spec.CredentialsSecret != nil {
+		credentialsSecretName = awsPlacementGroup.Spec.CredentialsSecret.Name
+	}
+
+	awsClient, err := awsclient.NewValidatedClient(
+		r.Client, credentialsSecretName, awsPlacementGroup.Namespace,
+		infra.Status.PlatformStatus.AWS.Region, r.ConfigManagedClient)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if r.AWSClient == nil {
+		r.AWSClient = awsClient
+	}
```

Comment:
I wrote this like so because in testing I'm overriding the AWS Client with a gomock'd version of it. Like so:
```
		mockCtrl = gomock.NewController(GinkgoT())
		mockAWSClient = mockaws.NewMockClient(mockCtrl)
		// Set up the reconciler.
		reconciler := &AWSPlacementGroupReconciler{
			Client:              mgr.GetClient(),
			Log:                 log.Log,
			ConfigManagedClient: mgr.GetClient(),
			AWSClient:           mockAWSClient,
		}
```

Which is then used in the controller and set up with the manager.

That said if there are better ways to do so while also retaining the ability to override the client at a reconcile/manager level I'm happy to change it :)
WDYT? Thanks.

Reply:
IIRC the way we do this in the MAPA actuator is to have a constructor function, perhaps we should use the same design there?

In the test you set your mock aws builder, in the real code, it uses a real aws client builder

Reply:
That's a good idea!
Implemented in: https://github.com/openshift/machine-api-provider-aws/pull/33/commits/386ff82f367040fbbf4cdce0c1e73448963f3ff2
LMK.